pythran/analyses/range_values.py

```diff
@@ -0,0 +1,448 @@
+        """ Set default range value for globals and attributs.
+
+        >>> import ast
+        >>> from pythran import passmanager, backend
+        >>> node = ast.parse("def foo(a, b): pass")
+        >>> pm = passmanager.PassManager("test")
+        >>> res = pm.gather(RangeValues, node)
+        >>> res['a']
+        Range(low=-inf, high=inf)
+        """
+        for global_name in self.globals:
+            self.result[global_name] = self.UNKNOWN_RANGE
+        for attr in node.args.args:
+            self.result[attr.id] = self.UNKNOWN_RANGE
+
+        map(self.visit, node.body)
+
+    def visit_Assign(self, node):
+        """
+        Set range value for assigned variable.
+
+        We do not handle container values.
+
+        >>> import ast
+        >>> from pythran import passmanager, backend
+        >>> node = ast.parse("def foo(): a = b = 2")
+        >>> pm = passmanager.PassManager("test")
+        >>> res = pm.gather(RangeValues, node)
+        >>> res['a']
+        Range(low=2, high=2)
+        >>> res['b']
+        Range(low=2, high=2)
+        """
+        low, high = self.visit(node.value)
+        for target in node.targets:
+            if isinstance(target, ast.Name):
+                self.add(target.id, low, high)
+
+    def visit_AugAssign(self, node):
+        """ Update range value for augassigned variables.
+
+        >>> import ast
+        >>> from pythran import passmanager, backend
+        >>> node = ast.parse("def foo(): a = 2; a -= 1")
+        >>> pm = passmanager.PassManager("test")
+        >>> res = pm.gather(RangeValues, node)
+        >>> res['a']
+        Range(low=1, high=2)
+        """
+        if isinstance(node.target, ast.Name):
+            low, high = self.combine(self.result[node.target.id],
+                                     self.visit(node.value),
+                                     node.op)
+            self.add(node.target.id, low, high)
+
+    def visit_For(self, node):
+        """ Handle iterate variable in for loops.
+
+        >>> import ast
+        >>> from pythran import passmanager, backend
+        >>> node = ast.parse('''
+        ... def foo():
+        ...     a = b = c = 2
+        ...     for i in range(1):
+        ...         a -= 1
+        ...         b += 1''')
+        >>> pm = passmanager.PassManager("test")
+        >>> res = pm.gather(RangeValues, node)
+        >>> res['a']
+        Range(low=-inf, high=2)
+        >>> res['b']
+        Range(low=2, high=inf)
+        >>> res['c']
+        Range(low=2, high=2)
+        """
+        assert isinstance(node.target, ast.Name), "For apply on variables."
+        self.add(node.target.id, *self.visit(node.iter))
+        old_range = copy.deepcopy(self.result)
+        map(self.visit, node.body)
+        for name, (low, high) in old_range.iteritems():
+            if self.result[name].low < low:
+                self.result[name] = \
+                    self.result[name]._replace(low=-float("inf"))
```

Comment:
:) namedtuple is already a good improvement compare to tuple we have in others part of the code :) but it will appear soon ;)